### ISSUE TITLE
platform specific intrinsics

### ISSUE THREAD

**Issue by oli-obk** [rust-rfcs/const-eval#7]
There are multiple thousand intrinsics for various platforms which users might want to use to speed up their code. It is unrealistic to implement all of these (although a certain subgroup like https://github.com/rust-lang-nursery/stdsimd/blob/05c2f61c384e2097a3a4c648344114fc4ac983be/coresimd/simd_llvm.rs seem to be manageable.

@rkruppe and @gnzlbg suggested a way to "runtime" detect whether we're in const eval or runtime (which would get completely optimized during codegen). While this would be trivial to implement, it seems very unfortunate to have to resolve to this.
There's precedent in c++ for such an escape hatch.

I am not very happy that we even need that many intrinsics outside the libstd and don't quite understand why these can't be optimizations that "just happen". Requiring users to "fix" optimizer deficiencies by giving them multiple thousand ways to gain a few cycles seems suboptimal.

It's perfectly fine not to support calling e.g. an interrupt register fiddling intrinsic at compile-time (there's no use for doing that, or it could be a noop in many cases). But when/if performance intrinsics are sprinkled onto code used by many users and there's an algorithm that works both at compile-time and runtime and could be optimized, it seems very unfortunate to scrap compile-time evaluability.

**Comment by rkruppe** [comment]
For the record, I did not advocate a "runtime" way for user code to behave differently in const eval, I am probably as uncomfortable with it as you are.

**Comment by gnzlbg** [comment]
There is a solution proposed for C++20 (I am not sure if it has been merged) called `is_constant_evaluated` (latest paper: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0595r1.html) that attacks this problem. 

Having to choose between whether code can be const evaluated or whether it performs good at run-time is a weird spot to be in as a programmer. I don't know whether this problem is worth solving, but the problem is real.

**Comment by oli-obk** [comment]
@gnzlbg gave us a great summary on irc. Here's the gist:

* while it moved from LEWG to CWG and LWG, people had mixed feelings about it
  * basically, std library writers want it,
  * and compiler writers wanted to expose it as an extension
* everybody agreed that it shouldn't be a widely used feature, and that it is hard to reason about it
* it was also argued when exactly should it return true
    * const FOO: i32 = foo(); it will return true
    * but what about let foo: i32 = foo(); ? 
    * should it also return true if `foo()` is const evaluated ?
    * will that depend on compiler-optimizations ?
      * so there was some agreement that it should not, and that it should be clear what it returns for every program
* they changed the name from `if constexpr() { ... }` to if std::is_constant_evaluated() to discourage its use
* clang/LLVM main devs argued strongly on its favour

**Comment by oli-obk** [comment]
Note that we can "just" add such a `std::is_constant_evaluated` function and make it `unsafe` to make it clear what's going on. The unsafe documentation would then state that the result of this function may not change the behaviour of any code. There could even be an additional `-Z` flag for fuzzers and testing that turns the flag to `true` during regular compilation

**Comment by gnzlbg** [comment]
When `miri` evaluates Rust program like this:

```rust
fn bar(x: i32) -> i32 { x * 2 }
const fn foo(x: i32) -> i32 {
    if unsafe { is_constant_evaluated() } {
        x * 2
    } else {
        // unsafe required because `bar` is 
        // not a `const fn` (hence not `const`-safe):
        unsafe { bar(x) }
    }
}
```

it could attempt to evaluate both branches, and compare that both produce the same result, which is the only effect that both branches can have. If the `bar` branch attempts to mutate global memory, read from a `static`, etc. `miri` could error.

**Comment by oli-obk** [comment]
We have changed course. Discussion is in https://github.com/rust-lang/rust/pull/64683/files/223c832b3894fe6ce6d61d4f459f0aa827bec264#discussion_r327153517

The TLDR is that we'd create an intrinsic

```rust
pub fn const_eval_select<ARG, F, G, RET>(arg: ARG, called_in_const: F, called_at_rt: G) -> RET;
```

that calls `called_in_const(arg)` if we're in a const eval context and `called_at_rt(arg)` if not. `F` and `G` must be function types (not function pointers).

**Comment by minerva** [comment]
> When `miri` evaluates Rust program like this:
> 
> ```rust
> fn bar(x: i32) -> i32 { x * 2 }
> const fn foo(x: i32) -> i32 {
>     if unsafe { is_constant_evaluated() } {
>         x * 2
>     } else {
>         // unsafe required because `bar` is 
>         // not a `const fn` (hence not `const`-safe):
>         unsafe { bar(x) }
>     }
> }
> ```
> 
> it could attempt to evaluate both branches, and compare that both produce the same result, which is the only effect that both branches can have. If the `bar` branch attempts to mutate global memory, read from a `static`, etc. `miri` could error.

How about a language-level `if const <expr> { .. }`? I think C++20 demonstrates why a library function to handle this job in conjunction with a regular `if` is a bad solution.

```rust
const fn foo(x: i32) -> i32 {
    if const {
        x * 2
    } else {
        unsafe { bar(x) }
    }
}
```

Similar how ISO NB comments tried to change C++20's `if (std::is_constant_evaluated()) { .. }` to
`if consteval (<expr>) { .. }`.

**Comment by oli-obk** [comment]
I think we race-conditioned on posting. See my above post for a solution that is much simpler in all cases, since it doesn't pollute the const part of a function with the non-const part.

**Comment by tarcieri** [comment]
Regarding `const` and `asm!`, I'm curious if there might be some way to integrate `const_eval_select` into CPU feature detection macros like `is_x86_feature_detected!`.

In the cases where we (prospectively) want to use `asm!`, it's accompanied with CPU feature detection with a pure Rust fallback anyway, so I'm curious if it would be possible to integrate these features together such that `asm!` could be allowable in a `const fn` if it's always guarded by a CPU feature detection macro.

**Comment by oli-obk** [comment]
> Regarding `const` and `asm!`, I'm curious if there might be some way to integrate `const_eval_select` into CPU feature detection macros like `is_x86_feature_detected!`.

Not exactly. The problem is that you cannot have non-const and const code within the same function, you need to have separate functions, so what I would imagine would work is

```rust
x86_feature_select!(the_feature, non_const_fn, const_fn, (first_arg: Foo, second_arg:Bar) ->ReturnType)
```

which expands to

```rust
fn foobar(first_arg: Foo, second_arg: Bar) -> ReturnType {
    if is_x86_feature_detected!(the_feature) {
        non_const_fn(first_arg, second_arg)
    } else {
        const_fn(first_arg, second_arg)
    }
}
const_eval_select((first_arg, second_arg), const_fn, foobar)
```

This should work right now.

**Comment by tarcieri** [comment]
I noticed `const_eval_select` isn't yet callable as a `const fn`. Is the plan to eventually allow that? (which would violate the rules for a `const fn` calling a non-`const fn`, but that would never actually happen in a `const fn` context)

**Comment by RalfJung** [comment]
I don't quite know what you mean? The standard library calls `const_eval_select` in `const fn` in quite a few places.

**Comment by tarcieri** [comment]
So it's callable from a `const fn` even though it's not declared a `const fn` itself? I see it defined as:

https://github.com/rust-lang/rust/blob/2862500152d89996ab2b682f22ed7e8923396992/library/core/src/intrinsics.rs#L2499

```rust
pub fn const_eval_select<ARG: Tuple, F, G, RET>(
```

**Comment by RalfJung** [comment]
It's a super special magic intrinsic. And highly unstable. ;)

**Comment by RalfJung** [comment]
Also, intrinsics are never `const fn`. That's not even accepted by rustc. You can't declare an `extern` fn to be `const`, in most situations that wouldn't make any sense. They care callable from `const fn` if they have a `rustc_const_(un)stable` attribute.

**Comment by tarcieri** [comment]
Any hopes for a stable API for this sort of thing in the future?

As an example use case beyond SIMD: I would like to use `asm!` to invoke `cmov*`/`csel*` for constant-time predication when available, but fall back on e.g. a bitmask-based selection in `const fn`, which seems easy enough to write with `const_eval_select`.

In `crypto-bigint` we ended up adding an entire `const fn`-friendly `subtle`-alike system for constant-time selection in order to solve this problem, which we provide in addition to the `subtle` traits. That approach comes with worries that LLVM's optimizer might rewrite bitmasking with branches in the non-`const` case, which is what we'd otherwise use `subtle`'s optimization barriers to avoid.

**Comment by RalfJung** [comment]
Hopes? Yeah for sure.^^ We even had an RFC that was meant to pave the way: https://github.com/rust-lang/rfcs/pull/3352. However, that didn't go anywhere, so a new approach is needed. I don't think currently anyone is pushing for it. Personally I'm considering const_mut_refs to be higher priority so that's the big const-eval feature I am currently pushing towards.

The current interface is also rather awkward, with how it forces you to write two *functions* to select between. It's really just a hack, and we'd probably want something nicer.
